src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -51,11 +53,17 @@ internal SocketConnection(Socket socket, SocketTransport transport)
                 Task receiveTask = DoReceive();
                 Task sendTask = DoSend();
 
-                // Wait for eiher of them to complete (note they won't throw exceptions)
-                await Task.WhenAny(receiveTask, sendTask);
+                // If the sending task completes then close the receive
+                // We don't need to do this in the other direction because the kestrel
+                // will trigger the output closing once the input is complete.
+                if (await Task.WhenAny(receiveTask, sendTask) == sendTask)
+                {
+                    // Tell the reader it's being aborted
+                    Volatile.Write(ref _aborted, true);
```

Comment:
One alternative would be to dispose the socket here and catch that exception in the reader but that seems grosser.